File: src/main/java/mt/edu/um/cf2/jgribx/grib2/ProductDiscipline.java
package mt.edu.um.cf2.jgribx.grib2;

import java.util.*;

public class ProductDiscipline
{
    private final int value;
    private final String name;
    private final List<ParameterCategory> categories;

    private static final Map<Integer, String> entries = new HashMap<>();

    static {
        entries.put(0, "METEOROLOGICAL");
        entries.put(1, "HYDROLOGICAL");
        entries.put(2, "LAND_SURFACE");
        entries.put(3, "SATELLITE_REMOTE_SENSING");
        entries.put(10, "OCEANOGRAPHIC");
    }

    public ProductDiscipline(int discipline)
    {
        name = entries.get(discipline);
        value = discipline;
        categories = ParameterCategory.getCategories(discipline);
    }

    public List<ParameterCategory> getParameterCategories()
    {
        return categories;
    }

    public static List<ProductDiscipline> getValues()
    {
        List<ProductDiscipline> list = new ArrayList<>();
        for (Integer key : entries.keySet())
        {
            list.add(new ProductDiscipline(key));
        }
        return list;
    }

    @Override
    public boolean equals(Object obj)
    {
        if (obj == null || !ProductDiscipline.class.isAssignableFrom(obj.getClass()))
        {
            return false;
        }
        return value == ((ProductDiscipline) obj).value;
    }

    @Override
    public String toString()
    {
        return entries.get(value);
    }
}

File: src/main/java/mt/edu/um/cf2/jgribx/grib2/ProductDiscipline.kt
package mt.edu.um.cf2.jgribx.grib2

import mt.edu.um.cf2.jgribx.grib2.ParameterCategory.Companion.getCategories

class ProductDiscipline private constructor(internal val value: Int, private val name: String = "") {
	companion object {
		val MISSING = 255
		val VALUES = mapOf(
				0 to "METEOROLOGICAL",
				1 to "HYDROLOGICAL",
				2 to "LAND_SURFACE",
				3 to "SATELLITE_REMOTE_SENSING",
				10 to "OCEANOGRAPHIC")
				.map { (k, v) -> ProductDiscipline(k, v) }
	}

	val parameterCategories = getCategories(value)

	override fun equals(other: Any?): Boolean = this === other
			|| other is ProductDiscipline
			&& value == other.value

	override fun hashCode(): Int = value

	override fun toString(): String = name
}
